### DISCUSSION TITLE
Should LazyVim migrate to nvim-treesitter main branch?

### DISCUSSION THREAD

**Discussion by reegnz** [LazyVim/LazyVim#3045] [Ideas]
The current default branch `master` is soft-frozen, the developers say `main` is where active development is happening right now.
Should LazyVim start migrating to that codebase?

https://github.com/nvim-treesitter/nvim-treesitter/discussions/6220#discussioncomment-8630595

**Comment by dpetka2001** [comment]
The `main` branch will become the default one with the release of Neovim 0.10, so is there an immediate need to rush this?

**Reply by reegnz** [reply]
No immediate need, but control when to break end users would be with lazyvim instead of nvim-treesitter if lazyvim would at least pin to `master`, no matter what the default branch is currently.
That would allow migration to the `main` branch happen on lazyvim's terms and timeline, because I assume `main` will have breaking changes compared to `master`.

**Reply by dpetka2001** [reply]
I'm assuming that since `main` will become default with the release of Neovim 0.10, there shouldn't be any significant problems transitioning. Except if `nvim-treesitter` changes its plan. I'm sure there will be some kind of announcement with the release of Neovim 0.10, so better wait until then. In the meantime nothing changes in LazyVim's behavior I would assume.

PS: Obviously the one more appropriate to give you a definitive answer is the maintainer himself, so let's just wait until he's back.

**Reply by FlippingBinary** [reply]
@dpetka2001 They still have not made `main` the default branch, but LazyVim will break when that happens unless changes are made to `lua/lazyvim/plugins/treesitter.lua`. Their readme says it's not compatible with lazy loading and the new setup method takes only an optional path to override the default installation location. Installing languages is done with the `install` method. Also, `nvim-treesitter.configs` doesn't exist anymore, and neither does `nvim-treesitter.query_predicates`.

**Reply by dpetka2001** [reply]
Maintainer is more appropriate for the feedback you're looking for.

**Reply by dpetka2001** [reply]
@FlippingBinary Maintainer had tried to make it available as an Extra at some point, but removed it at the request of a Neovim core maintainer. See https://github.com/LazyVim/LazyVim/pull/3177

I'm sure when he has input that it's more stable, he will re-visit the idea either of a separate Extra like before or incorporating into default. 
There's also a separate branch [treesitter-rewrite](https://github.com/LazyVim/LazyVim/compare/main...treesitter_rewrite) from that point at time.

**Reply by FlippingBinary** [reply]
Thank you for the links! I'll give it another try. Even though `main` still isn't their default branch, they are closing issues related to the `master` branch as unplanned and telling people to use `main`. I'm not sure if that means they would be okay with LazyVim switching over, but I don't want to get caught off guard at a bad time, lol. Plus, the new rewrite looks cleaner and might fix some issues I have in some environments.

**Reply by dusty-phillips** [reply]
I think this needs to be revisited now, though. The PR dpetka2001 mention was indeed closed a year ago because nvim-treesitter was not ready for the main branch to be default. The current advice as of two weeks ago on the nvim-treesitter Readme is to prefer the main branch. As FlippingBinary mentioned, this is a breaking change.

**Reply by dpetka2001** [reply]
I created a custom user Extra with the code from the aforementioned branch and 2 additions of my own to make things work.

You can find the diff [here](https://github.com/dpetka2001/nvim_treesitter_main/commit/8f63131565a66e7402964497cee021359d99e08b). 

This is for anyone interested in testing the `nvim-treesitter` main branch in their personal configuration. I hope I didn't miss anything.

Of course ideally this will be handled by Folke himself when he comes back, but until then you can test things out on your own if you're interested.

**Reply by dusty-phillips** [reply]
Confirmed working. Thanks @dpetka2001! I can now use the new gleam assert syntax in my tests.

**Reply by rmunn** [reply]
> Of course ideally this will be handled by Folke himself when he comes back, but until then you can test things out on your own if you're interested.

Sorry for the off-topic question, @dpetka2001, but I can't find Folke's announcement of being away that you seem to be referring to. I vaguely remember reading it at some point but I can't find it anymore (and for all I know, I may be getting mixed up with a different maintainer of a different project). Do you know where it is, and/or when he said he'd be back? I'm interested in trying out nvim-treesitter main branch myself, but I don't want to put in the work of customizing the extra if Folke is about to do it "officially" next week or something. Do you have a link to where he said how long he expects to be gone? (Or is it an indefinite, "until I have more time to work on LazyVim" kind of thing?)

**Reply by dpetka2001** [reply]
There was no announcement. He randomly commented on a Reddit post [here](https://www.reddit.com/r/neovim/comments/1kj5w8q/is_there_any_distribution_such_as_lazyvim_but/mrkrtdf/). In that comment he mentions that he'll be travelling for most of July-September. 

So, I would expect that he would be back at some time after October comes. Except if other plans come into place or he just stops doing Neovim stuff all together. Who knows? 

Regarding nvim-treesitter main branch I linked a diff in my previous comment. You can just copy/paste it in your personal configuration and then enable the User Extra from the `:LazyExtras` UI. I've been using it without any problems so far.

**Comment by dusty-phillips** [comment]
I was struggling with nvim-treesitter-textobjs and kulala.nvim not working because of the patch in dpetka2001's file. Kulala has detection for both versions of treesitter, but it uses the existence of `get_parser_configs` to decide which one to use. The patch was causing it to recognize master when it should be main.

My fix was to remove the patch function from dpetka's config which looks like [this](https://github.com/dusty-phillips/dotfiles/blob/main/.config/nvim/lua/plugins/extras/ui/treesitter-rewrite.lua). This fixed kulala.nvim, but caused a failure inside nvim-treesitter-textobjects. Turns out that project also has a "master/main" split and they've completely changed the API on main. My config overrides the lazymin `nvim-treesitter-textobjects` config and looks like [this](https://github.com/dusty-phillips/dotfiles/blob/main/.config/nvim/lua/plugins/extend-nvim-treesitter-textobjects.lua).
